### PR TITLE
chore: add script to update '@opentelemetry/*' deps in this repo

### PR DESCRIPTION
**Pull request by trentm**
Refs: https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1917

* * *

This is a replacement for my earlier PR https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1974 where I had mistakenly created a PR from the *main* branch of my fork. This corrects that to use a feature branch. Sorry for that shuffle.

### CODE REVIEW

File: scripts/update-otel-deps.js [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+
+function getAllWorkspaceDirs() {
+    const pj = JSON.parse(
+        fs.readFileSync(path.join(TOP, 'package.json'), 'utf8')
+    );
+    const workspaceDirs = pj.workspaces
+        .map((wsGlob) => globSync(path.join(wsGlob, 'package.json')))
+        .flat()
+        .map(path.dirname);
+    return workspaceDirs;
```
[the remark this answers is not in the document]
Comment (trentm) [reply scripts/update-otel-deps.js]: Done in commit 4092119e0167c1587c2cb3bd28e0ee1a43f96da2.

File: scripts/update-otel-deps.js [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        };
+    }
+    debug('pkgInfoFromWsDir: ', pkgInfoFromWsDir);
+
+    console.log('\nGathering info on outdated deps in each workspace:');
+    const summaryStrs = new Set();
+    const npmInstallTasks = [];
+    const npmUpdatePkgNames = new Set();
+    let i = 0;
+    for (let wsDir of wsDirs) {
```
[the remark this answers is not in the document]
Comment (trentm) [reply scripts/update-otel-deps.js]: Done in commit 4092119e0167c1587c2cb3bd28e0ee1a43f96da2.

File: scripts/update-otel-deps.js [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+ *      accurately represent the commands run, because the final 'npm update'
+ *      args can depend on the results of earlier 'npm install' commands.
+ */
+function updateNpmWorkspacesDeps({patterns, allowRangeBumpFor0x, dryRun}) {
+    assert(
+        patterns && patterns.length > 0,
+        'must provide one or more patterns'
+    );
+
+    let p;
```
[the remark this answers is not in the document]
Comment (trentm) [reply scripts/update-otel-deps.js]: Done in commit 4092119e0167c1587c2cb3bd28e0ee1a43f96da2.

File: scripts/update-otel-deps.js [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+        // where that "opentelemetry-node" is a workspace dir *basename* (sigh).
+        p.stdout
+            .trim()
+            .split('\n')
+            .forEach((line) => {
+                const dependedBy = line.split(':')[4];
+                if (wsDirBasenames.has(dependedBy)) {
+                    return;
+                }
+                // TODO Do we want to guard this on `patterns`?
```
[the remark this answers is not in the document]
Comment (trentm) [reply scripts/update-otel-deps.js]: Yes, for this usage (where `patterns` is hardcoded to `@opentelemetry/*`) it should always be fine. If the script was generalized a bit to allow passing in a pattern that only matched a subset of the packages in the workspace -- e.g. `@opentelemetry/instrumentation-*` or something -- then it *might* be something to consider.

I'm happy to remove a TODO. I've dropped it in commit 4092119e0167c1587c2cb3bd28e0ee1a43f96da2.

File: scripts/update-otel-deps.js [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+                err.cwd = task.cwd;
+                err.argv = task.argv;
+                err.process = p;
+                throw err;
+            }
+            // Note: As noted at https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1917#issue-2109198809
+            // (see "... requires running npm install twice ...") in some cases this
+            // 'npm install' doesn't actually install the new version, but do not
+            // error out!
+            // TODO: guard against this with 'npm ls' or package.json check?
```
[the remark this answers is not in the document]
Comment (trentm) [reply scripts/update-otel-deps.js:225]: > Though we'll likely not run into it anymore now that we've moved the pacakge that caused this to the incubator directory,

Unfortunately I think the weird case can still come up. David hit it recently in TAV tests with interplay between the 'mongodb' version being tested for TAV tests in instrumentation-mongodb and the pinned version of mongodb as a transitive dep of instrumentation-mongoose. See https://github.com/open-telemetry/opentelemetry-js-contrib/pull/2001#issuecomment-1992202942
Yuck.

[the remark this answers is not in the document]
Comment (trentm) [reply scripts/update-otel-deps.js:225]: Do you prefer to have an issue created that might languish? or just leave the TODO in-place?

**Comment by trentm** [comment]
Testing: I've run the script a couple times after the above updates and it still seems to work correctly.

File: scripts/update-otel-deps.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (pichlermarc) [reply scripts/update-otel-deps.js:225]: Let's just leave the TODO in place. :+1:

**Review by pichlermarc — APPROVED** [review]
Thanks again for working on this.  :+1: 
This was immensely helpful during the last release. :slightly_smiling_face:

**Pull request merged**